src/sdk/util/ssi/vc.ts

```diff
@@ -0,0 +1,371 @@
+     *
+     * @returns {Promise<VerifiableCredential>} the Verifiable Credential
+     */
+    static async sign<T extends object = object>(
+        id: DIDurl,
+        type: string[],
+        credentialSubject: T,
+        issuer: Issuer,
+        options: {
+            subject?: URL;
+        } = {}
+    ): Promise<VerifiableCredential<T>> {
+        const vc: W3CCredential = {
+            '@context': ['https://www.w3.org/2018/credentials/v1'],
+            id,
+            type,
+            issuer: {
+                id: issuer.did,
+            },
+            issuanceDate: new Date().toISOString(),
+            credentialSubject: credentialSubject as any,
+        };
+
+        if (options.subject) vc.credentialSubject.id = options.subject;
+
+        const jwt = await createVerifiableCredentialJwt(vc, issuer, { canonicalize: true });
+
+        return new VerifiableCredential<T>(jwt);
+    }
+
+    /**
+     * Returns the JWT string
+     *
+     * @returns {string} the JWT string
+     */
+    toString(): string {
+        return this.jwt;
+    }
+
+    /**
+     * Returns the JWT string, called by JSON.stringify
+     *
+     * @returns {string} the JWT string
+     */
+    toJSON(): string {
+        return this.toString();
+    }
+}
+
+export type VerifiableCredentialOptions = {
+    subject?: URL;
+    additionalTypes?: string[];
+};
+
+export type VCWithTypeType<T extends object = object> =
+    | VerifiableCredential<{ payload: T; type: string }>
+    | VerifiableCredentialWithType<T>
+    | JWT;
+
+/**
+ * A wrapper that adds a type object to VCs to allow for identification, and presents a simper interface
+ *
+ * This is the base class. It is expected that extension classes will be created for the different VC types.
+ *
+ * @inheritdoc sign() is a protected alternative constructor. In child classes, a new alternative constructor
+ * should be created which returns the child class type.
+ * @inheritdoc if the payload type requires ad-hoc decoding, override the constructor and decode the payload
+ *
+ * @example see an example of the above in the LoginRequestMessage class in `../..services/communication/messages.ts`
+ */
+export class VerifiableCredentialWithType<T extends object = object> implements Serializable {
+    private vc: VerifiableCredential<{ payload: T; type: string }>;
+    protected decodedPayload: T;
+
+    /**
+     *
+     * @param vc the VC to wrap
+     *
+     * @inheritdoc override me if the payload type requires ad-hoc decoding
+     */
+    constructor(vc: VCWithTypeType<T>) {
+        if (typeof vc === 'string') {
+            this.vc = new VerifiableCredential<{ payload: T; type: string }>(vc);
+        } else if (vc instanceof VerifiableCredentialWithType) {
+            this.vc = vc.getVc() as VerifiableCredential<{ payload: T; type: string }>;
+        } else {
+            this.vc = vc;
+        }
+
+        this.decodedPayload = this.getVc().getCredentialSubject().payload;
+    }
```

Comment:
why creating a new class for this cant be an optional property ?

Reply:
talked with Rebal and accepted